text/0000-ide.md

```diff
@@ -0,0 +1,530 @@
+shared, memory mapped files as a buffer; I'm not familiar enough with the
+library to work out what is needed).
+
+I've detailed the API I believe we'll need to start with. This is slightly more
+than a minimal set. I expect it will expand as time goes by. At some point we
+will want to stabilise parts of the API to allow for third party implementations
+of the oracle and compiler.
+
+All API calls can return success or error results. Many calls involve a *span*;
+for the oracle's API, this is defined as two byte offsets from the start of the
+file (oracle spans must always be contained in a single file).
+
+There are some alternative span definitions: we could use file and column indices
+rather than byte offsets (this has some edge case difficulties with the
+definition of a newline - do unicode newlines count? It also requires some extra
+computation), we could use character offsets (again involves some more
+computation, but might be more robust).
+
+A problem is that Visual Studio uses UTF16 while Rust uses UTF8, there is (I
+understand) no efficient way to convert between byte counts in these systems.
+I'm not sure how to address this. It might require the oracle to be able to
+operate in UTF16 mode.
+
+Where no return value is specified, the call returns success or failure (with a
+reason).
+
+The philosophy of the API is that most functions should only take a single call,
+as opposed to making each function as minimal and orthogonal as possible. This
+is because IPC can be slow and response time is important for IDEs.
+
+
+**Projects**
+
+Note that the oracle stores no metadata about a project.
+
+*init project*
+
+Takes a project name, returns an id string (something close to the project's name).
+
+*delete project*
+
+Takes a project id.
+
+*list projects*
+
+Takes nothing, returns a list of project ids.
+
+Each of the remaining calls takes a project identifier.
+
+
+**Update**
+
+See section on input data format below.
+
+*update*
+
+Takes input data (actual source code rather than spans since we cannot assume
+the user has saved the file) and a list of spans to invalidate.  Where there are
+no invalidated spans, the update call adds data (which will cause an error if
+there are conflicts). Where there is no input data, update just invalidates.
+
+We might want to allow some shortcuts to invalidate an entire file or
+recursively invalidate a directory.
+
+
+**Description**
+
+*get definition*
+
+Takes a span, returns all 'definitions and declarations' for the identifier
+covered by the span. Can return an error if the span does not cover exactly one
+identifier or the oracle has no data for an identifier.
+
+The returned data is a list of 'defintion' data. That data includes the span for
+the item, any documentation for the item, a code snippet for the item,
+optionally a type for the item, and one or more kinds of definition (e.g.,
+'variable definition', 'field definition', 'function declaration').
```

Comment:
Could you explain the push-driven update a bit more please? I am assuming the IDE plugin is pretty much state-less with respect to the oracle's data and whenever the user performs an action (right click, hover, hotkey, etc.) then it will query the oracle for the data it needs (thus why the oracle must be quick).